rir/src/compiler/analysis/loop_detection.h

```diff
@@ -63,6 +66,35 @@ class LoopDetection {
             o->isInnermost_ = false;
         }
 
+        /*
+         * TODO: finds a preheader when there is one.
+         * We should create a preheader when there is non.
+         */
+        BB* preheader(const CFG& cfg) {
+            std::vector<BB*> outOfLoopPredecessor;
+            for (auto pred : cfg.immediatePredecessors(header())) {
+                if (!body_.count(pred))
+                    outOfLoopPredecessor.push_back(pred);
+            }
+
+            return outOfLoopPredecessor.size() != 1
+                       ? nullptr
+                       : outOfLoopPredecessor.front();
+        }
+
+        bool holdsPropery(const InstrActionPredicate& action) {
+            auto answer = true;
```

Comment:
My personal habit is to write out types, unless they're too long or it's redundant.